theano/compile/function_module.py

```diff
@@ -483,6 +488,66 @@ def __copy__(self):
                 there.data = here.data
         return cpy
 
+    def copy(self, share_memory = False):
+        """
+        Copy this function. Copied function will have separated maker and fgraph
+        with original function. User can choose whether to separate storage by 
+        changing the share_memory arguments
+        ---------------------
+        Params:
+            share_memory -- { boolean } Default is False. When True, two function
+            share intermediate storages(storages except input and output storages)
+        ---------------------
+        Returns:
+            func -- Copied theano.Function
+        """
+
+        if not share_memory:
+            return self.__copy__()
+        else:
+            # copy SymbolocKits
+            ins, outs = copy.deepcopy([self.maker.inputs, self.maker.outputs])
```

Comment:
I am not sure about this. Initially I thought that variables in ```In()``` instances share same references with those in fgraph. However, it seem that it's not the case.
```
In [37]: func = theano.function([x],[x+1])
In [38]: ins = func.maker.inputs
In [39]: for i in ins:
    print i.variable in func.maker.fgraph.variables
   ....:     
False
```
The main problem is that matching two variable isn't easy, as now variables in ```In()```s aren't those in ```fgraph```, so memo mapping can't work. What's worse, as expression```new_var == old_var``` always return ```False``` even two var instances represent same things(at lease this happened in my experiments), I can hardly match variable in ```In()```s with those in fgraph. Also, after fgraph is copied, the order of variable is not guaranteed(right?), so I cannot loop over pairs. Apart from this, matching updates will also become a problem.  

All these problems occur only when variable in ```In()```s and fgraph need to share same reference, otherwise it will be easy. I don't even need to copy a variable.  

---------------------------------

Apart from this, I would like to know how ```In()``` affect the evaluation of function, like ```update```. Because I cannot find any relevant codes.